From e900c99db9296a268df9c3d0af4f01f6cf5f5e46 Mon Sep 17 00:00:00 2001
From: Emergi Idan <idan.emergi@intel.com>
Date: Tue, 6 Oct 2020 14:32:55 +0300
Subject: [PATCH] LMTSYS-4047: Fix/Change debugfs prints in LGM PPv4

Add total counter for reassembly counters
Fix init loop in hal buffer manager (bm)
add free CLIDs to rpb stats
---
 .../ppv4/pp-api/pp-session-mgr/pp_session_mgr_debugfs.c | 17 ++++++++++-------
 drivers/net/ppv4/pp-hal/buffer-manager/bm.c             | 12 ++++++------
 .../net/ppv4/pp-hal/receive-packet-buffer/rpb_debugfs.c | 10 +++++++++-
 3 files changed, 25 insertions(+), 14 deletions(-)

diff --git a/drivers/net/ppv4/pp-api/pp-session-mgr/pp_session_mgr_debugfs.c b/drivers/net/ppv4/pp-api/pp-session-mgr/pp_session_mgr_debugfs.c
index 8be714427f5c..6b9b837d1c51 100644
--- a/drivers/net/ppv4/pp-api/pp-session-mgr/pp_session_mgr_debugfs.c
+++ b/drivers/net/ppv4/pp-api/pp-session-mgr/pp_session_mgr_debugfs.c
@@ -2279,6 +2279,7 @@ s32 __smgr_reass_stats_show(char *buf, size_t sz, size_t *n, void *s, u32 num,
 {
 	struct reassembly_stats *stats, *it;
 	u32 cid;
+	u64 total;
 	char **str;
 	static const char * const cntrs_str[] = {
 		"rx_pkts",
@@ -2308,31 +2309,33 @@ s32 __smgr_reass_stats_show(char *buf, size_t sz, size_t *n, void *s, u32 num,
 
 	pr_buf(buf, sz, *n, "\n");
 	pr_buf_cat(buf, sz, *n,
-		   "|=======================================================================|\n");
+		   "|====================================================================================|\n");
 	pr_buf_cat(buf, sz, *n,
-		   "|                Reassembly UC Statistics                               |\n");
+		   "|                Reassembly UC Statistics                                            |\n");
 	pr_buf_cat(buf, sz, *n,
-		   "|=======================================================================|\n");
+		   "|====================================================================================|\n");
 	pr_buf_cat(buf, sz, *n,
 		   "| %-17s ", "Counter");
 	for (cid = 0; cid < num; cid++)
 		pr_buf_cat(buf, sz, *n, "|  CPU%u      ", cid);
-	pr_buf_cat(buf, sz, *n, "|\n");
+	pr_buf_cat(buf, sz, *n, "|  Total     |\n");
 	pr_buf_cat(buf, sz, *n,
-		   "|-------------------+------------+------------+------------+------------|\n");
+		   "|-------------------+------------+------------+------------+------------+------------|\n");
 
 	cid = 0;
 	stats = (struct reassembly_stats *)s;
 	for_each_arr_entry(str, cntrs_str, ARRAY_SIZE(cntrs_str), cid++) {
+		total = 0;
 		pr_buf_cat(buf, sz, *n, "| %-17s ", *str);
 		for_each_arr_entry(it, stats, num) {
 			pr_buf_cat(buf, sz, *n, "| %10llu ",
 				   *(((u64 *)it) + cid));
+			total += *(((u64 *)it) + cid);
 		}
-		pr_buf_cat(buf, sz, *n, "|\n");
+		pr_buf_cat(buf, sz, *n, "| %10llu |\n", total);
 	}
 	pr_buf_cat(buf, sz, *n,
-		   "|-----------------------------------------------------------------------|\n\n");
+		   "|------------------------------------------------------------------------------------|\n\n");
 
 	return 0;
 }
diff --git a/drivers/net/ppv4/pp-hal/buffer-manager/bm.c b/drivers/net/ppv4/pp-hal/buffer-manager/bm.c
index 7f918c00761a..1ae293c0bb34 100644
--- a/drivers/net/ppv4/pp-hal/buffer-manager/bm.c
+++ b/drivers/net/ppv4/pp-hal/buffer-manager/bm.c
@@ -88,8 +88,7 @@ s32 __bm_pool_num_bursts_set(u8 pool_id, u8 num_bursts)
 
 s32 bm_init(const struct pp_bm_init_param * const cfg)
 {
-	u32 reg_val = 0;
-	u8  pool_idx;
+	u32 idx, reg_val = 0;
 	s32 ret = 0;
 	u32 burst_sz, burst_sz_val, burst_num, burst_num_val;
 
@@ -125,12 +124,12 @@ s32 bm_init(const struct pp_bm_init_param * const cfg)
 
 	burst_sz = BM_OCPM_BURST_SIZE_256B;
 	burst_num = BM_OCPM_NUM_OF_BURSTS_2;
-	for (pool_idx = 0; pool_idx < db->max_pools; pool_idx++) {
-		ret = __bm_pool_burst_size_set(pool_idx, burst_sz);
+	for (idx = 0; idx < db->max_pools; idx++) {
+		ret = __bm_pool_burst_size_set(idx, burst_sz);
 		if (ret)
 			goto bm_init_done;
 
-		ret = __bm_pool_num_bursts_set(pool_idx, burst_num);
+		ret = __bm_pool_num_bursts_set(idx, burst_num);
 		if (ret)
 			goto bm_init_done;
 	}
@@ -143,7 +142,8 @@ s32 bm_init(const struct pp_bm_init_param * const cfg)
 		BM_BURST_BYTES_TO_CACHE_ROWS(burst_sz_val, burst_num_val);
 
 	/* Reset group reserved buffers */
-	PP_REG_WR32(PP_BM_GRP_RES_REG_IDX(0), 0);
+	for (idx = 0; idx < db->max_groups; idx++)
+		PP_REG_WR32(PP_BM_GRP_RES_REG_IDX(idx), 0);
 
 	pr_debug("pp bm was initialized\n");
 
diff --git a/drivers/net/ppv4/pp-hal/receive-packet-buffer/rpb_debugfs.c b/drivers/net/ppv4/pp-hal/receive-packet-buffer/rpb_debugfs.c
index c01021954f69..f9bd89260bd5 100644
--- a/drivers/net/ppv4/pp-hal/receive-packet-buffer/rpb_debugfs.c
+++ b/drivers/net/ppv4/pp-hal/receive-packet-buffer/rpb_debugfs.c
@@ -221,7 +221,7 @@ PP_DEFINE_DEBUGFS(cfg_show, __rpb_cfg_show, NULL);
 void __rpb_stats_show(struct seq_file *f)
 {
 	struct rpb_hw_stats stats;
-	u32 port, tc;
+	u32 port, tc, reg, free_clids, free_cc;
 	s32 ret = 0;
 	bool port_back_pres;
 
@@ -229,6 +229,10 @@ void __rpb_stats_show(struct seq_file *f)
 	if (unlikely(ret))
 		return;
 
+	reg = PP_REG_RD32(PP_RPB_GEN_DBG_CNT_REG);
+	free_clids = PP_FIELD_GET(PP_RPB_GEN_DBG_CNT_FREE_CLIDS_MSK, reg);
+	free_cc    = PP_FIELD_GET(PP_RPB_GEN_DBG_CNT_FREE_CC_MSK, reg);
+
 	seq_puts(f, "\n");
 	seq_puts(f, "   Receive Packet Buffer Statistics\n");
 	seq_puts(f, "|==========================================|\n");
@@ -236,6 +240,10 @@ void __rpb_stats_show(struct seq_file *f)
 		   stats.total_bytes_used);
 	seq_printf(f, "|  %-25s = %-10u  |\n", "Global Packets Used",
 		   stats.total_pkts_used);
+	seq_printf(f, "|  %-25s = %-10u  |\n", "Free CLIDs",
+		   free_clids);
+	seq_printf(f, "|  %-25s = %-10u  |\n", "Free CCs",
+		   free_cc);
 
 	seq_puts(f, "|");
 	RPB_FOR_EACH_PORT(port)
